admin/controllers/school.js

```diff
@@ -368,18 +372,83 @@ const getSelectPupilNotTakingCheck = async (req, res, next) => {
 
   let attendanceCodes
   let formData
+  let pupilsList
+  let htmlSortDirection = []
+  let arrowSortDirection = []
 
+  // Sorting
+  const sortField = req.params.sortField === undefined ? 'name' : req.params.sortField
+  const sortDirection = req.params.sortDirection === undefined ? 'asc' : req.params.sortDirection
+
+  let sortingDirection = [
+    {
+      'key': 'name',
+      'value': 'asc'
+    },
+    {
+      'key': 'reason',
+      'value': 'asc'
+    }
+  ]
+
+  // Markup links and arrows
+  sortingDirection.map((sd, index) => {
+    if (sd.key === sortField) {
+      htmlSortDirection[sd.key] = (sortDirection === 'asc' ? 'desc' : 'asc')
+      arrowSortDirection[sd.key] = (htmlSortDirection[sd.key] === 'asc' ? 'sort up' : 'sort')
+    } else {
+      htmlSortDirection[sd.key] = 'asc'
+      arrowSortDirection[sd.key] = 'sort'
+    }
+  })
+
+  // Get attendance code index
   try {
     attendanceCodes = await AttendanceCode.getAttendanceCodes().exec()
   } catch (error) {
     console.log('ERROR getting attendance codes', error)
     return next(error)
   }
 
+  // Get pupils for user' school
+  const pupils = await fetchSortedPupilsData(req.user.School, 'lastName', sortDirection)
+  pupilsList = await Promise.all(pupils.map(async (p) => {
+    if (p.attendanceCode !== undefined) {
+      let num = p.attendanceCode.code
+      try {
+        p.reason = attendanceCodes[num].reason
+      } catch (error) {
+      }
+    } else {
+      p.reason = 'N/A'
+    }
+    return p
+  })).catch((error) => next(error))
+
+  // Sorting by 'reason' needs to be done using .sort
+  if (sortField === 'reason') {
+    pupilsList = await pupilsList.sort((a, b) => {
+      if (a.reason === 'N/A') {
+        return 1
+      } else if (b.reason === 'N/A') {
+        return -1
+      } else if (a.reason === b.reason) {
+        return 0
+      } else if (sortDirection === 'asc') {
+        return a.reason < b.reason ? -1 : 1
+      } else {
+        return a.reason < b.reason ? 1 : -1
+      }
+    })
+  }
+
   return res.render('school/select-pupils-not-taking-check', {
     breadcrumbs: req.breadcrumbs(),
     attendanceCodes,
-    formData
+    formData,
+    pupilsList,
+    htmlSortDirection,
+    arrowSortDirection
```

Comment:
We need unit tests on the controller please.

Reply:
yes.  Can we please all ensure that at the very least the server side code is covered with tests.  I will be implementing coverage checks into travis soon.